src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs

```diff
@@ -47,14 +47,17 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
             // first we can make a small optimization:
             // If left is a constant then we already know whether it is null or not. If it is null then we 
             // can simply generate "right". If it is not null then we can simply generate
-            // MakeConversion(left).
+            // MakeConversion(left). This does not hold when the left is an unconstrained type parameter: at runtime,
+            // it can be either left or right depending on the runtime type of T
 
-            if (rewrittenLeft.IsDefaultValue())
+            var isUnconstrainedTypeParameter = rewrittenLeft.Type?.IsReferenceType == false && rewrittenLeft.Type?.IsValueType == false;
+
+            if (rewrittenLeft.IsDefaultValue() && !isUnconstrainedTypeParameter)
             {
                 return rewrittenRight;
             }
 
-            if (rewrittenLeft.ConstantValue != null)
+            if (rewrittenLeft.ConstantValue != null && !isUnconstrainedTypeParameter)
```

Comment:
>isUnconstrainedTypeParameter [](start = 56, length = 28)

Is `isUnconstrainedTypeParameter` ever true when `ConstantValue != null`?

Reply:
Yes, you can have `default(T) ?? t1`.

---
In reply to: [213727960](https://github.com/dotnet/roslyn/pull/29577#discussion_r213727960) [](ancestors = 213727960)

Reply:
Won't `null` show up as a constant value?

Reply:
The constant value of the null literal is `ConstantValue.Null`, which is not actually `null`.

---
In reply to: [213847956](https://github.com/dotnet/roslyn/pull/29577#discussion_r213847956) [](ancestors = 213847956)